---
title: "ngs_assignment_wk13"
author: "Xiaoyan Wen"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Next Generation Sequence Analysis Homework Week 13

In this exercise, you will conduct reference free k-mer profiling to estimate genome size, heterozygosity and repetitive DNA content from raw Illumina short read data using GenomeScope (Vurture et al. 2017). You will create k-mer profiles with Jellyfish on the Greene cluster and then upload the k-mer histogram to the GenomeScope website to estimate parameters of interest and generate plots for you to analyze.

#### About the data

Whole genome re-sequencing data were collected from a date palm sample and an Arabidopsis thaliana (thale cress) ecotype (Chi-2) from 1001genomes.org

<https://1001genomes.org/data/SLU/SLUHenning2014/releases/2014_01_28/strains/Chi-2>

Date palm and Arabidopsis were selected because they have very different levels of heterozygosity, genome size and genome repeat content.

#### Task 1: Understanding k-mer profiles

A k-mer profile is a table of all k-mers of a given length and their abundance in the reads. Below is a hypothetical 12 bp read. Use your understanding of a k-mer profile (See Week 13 pre-recorded videos) to manually create one for k-mers of length 3 ("3-mers").

ATGCATGCTGCA

Report the profile in the form of a two column table for your answer [ 1 point ].

```{r eval=FALSE}
## occurrences  k-mers  
## 1            3 (CAT,CTG,GCT)
## 2            2 (ATG,GCA)
## 3            1 (TGC)
```

#### Task 2: Generate k-mer profile with Jellyfish

The first step in conducting a kmer-based analysis of genome size, repeat content and heterozygosity with GenomeScope is to generate a kmer profile of the sample genome. Since the GenomeScope approach is often used to estimate key parameters for de novo genome assembly studies (e.g., genome size and heterozygosity), it is common to generate the k-mer profile on Illumina paired-end reads directly from the fastqs and then make decisions about de novo genome sequencing (e.g., platform, depth, software) based on the kmer-based analysis.

Jellyfish is a software for counting the number of occurences of k-mers and generating a k-mer profile. For a given k-mer length k, the output is the number of occurences of all observed k-mers of length k.

Below you will generate a k-mer profile for the plant Arabidopsis thaliana (Thale cress) genome using Jellyfish. The date palm sample has been processed for you by your instructor.

Begin by logging in to Greene and doing the following:

    srun --time=4:00:00 --mem=4GB --pty /bin/bash
    cd $SCRATCH
    mkdir ngs.week13
    cd ngs.week13

The fastqs for this assignment are located in the following directory:

`/scratch/work/courses/BI7653/hw13.2022`

The paired-end fastqs for the Arabidopsis thaliana sample are: chi2_r1.fq.gz chi2_r2.fq.gz

Concatenate the paired-end fastqs using the following (note: you will need to write the output file to your /scratch):

    cat <read1.gz> <read2.gz> > <sample>.gz
```{bash eval=FALSE}
cat /scratch/work/courses/BI7653/hw13.2022/chi2_r1.fq.gz \
/scratch/work/courses/BI7653/hw13.2022/chi2_r2.fq.gz \
> chi2.fq.gz
```

You should now have one gzipped fastq. Jellyfish requires unzipped fastqs so unzip the concatenated gzipped file using gunzip before proceeding.

Q2.1. GenomeScope requires the read length for the fastq input file. What is the read length for the concatenated fastq for Arabidopsis [ 1 point ]?

```{bash eval=FALSE}
gunzip chi2.fq.gz
awk '{if(NR%4==2) {count++; bases += length} } END{print bases/count}' chi2.fq
## 89.9976
```
```{r}
sprintf("The read length for the concatenated fastq for Arabidopsis is: %f", 89.9976)
```

The GenomeScope website provides instructions for how to run Jellyfish on the concatenated fastq file to produce inputs suitable for GenomeScope. Read the instructions at the following link (note: you do not need to install Jellyfish as it is already available on Greene)

<http://qb.cshl.edu/genomescope/>

Now create a slurm script that loads the Jellyfish module and run the commands in the GenomeScope "Instructions" section. The goal is to generate the k-mer count and histogram outputs for the concatenated Arabidopsis fastq file.

Execute your script when you are ready.

Q2.1 Copy your slurm script with all commands into your homework document for your answer [ 2 points ]

```{bash eval=FALSE}
#!/bin/bash
#
#SBATCH --nodes=1
#SBATCH --tasks-per-node=10
#SBATCH --cpus-per-task=1
#SBATCH --time=2:00:00
#SBATCH --mem=12GB
#SBATCH --job-name=kmercount_slurm
#SBATCH --mail-type=FAIL
#SBATCH --mail-user=xw2470@nyu.edu   
module purge

echo script begin: $(date)

module load jellyfish/2.3.0

# Count kmers
jellyfish count -C -m 21 -s 1000000000 -t 10 chi2.fq -o chi2_reads.jf

# export histogram
jellyfish histo -t 10 chi2_reads.jf > chi2_reads.histo

module purge
echo script completed: $(date)
```

#### Task 3: Reference-free estimation of heterozygosity, genome size and repeat content

You will now run GenomeScope on your histogram file produced by Jellyfish at the GenomeScope website. When your Jellyfish job is completed on Greene, download the .histo output to your personal computer. Then go to the GenomeScope website and upload the Jellyfish .histo file for the Arabidopsis sample.

Run GenomeScope taking care to include the correct read length and k-mer size (from your Jellyfish command).

To assist in answering the questions below, you may wish to watch the pre-recorded video on NYU Brightspace, read the GenomeScope publication (this week's assigned reading: Vurture et al. 2017), visit the "Examples" link on the GenomeScope website.

Q3.1 Report the settings (kmer length, read length, and max kmer coverage ) you input on the GenomeScope website [ 1 point ]. 
![](settings.png)


Q3.2. Report your GenomeScope outputs [ 1 point ] (1) Paste the GenomeScope Results table output for Arabidopsis into your homework document (2) Include the two kmer profile plots generated by GenomeScope. Download images or take a screen shot of the output plots and include in your report. 

![Result Table](result_table.png)


![Result Plots](result_plots.png)



Q3.3 Arabidopsis thaliana is an inbreeding species capable of self-fertilization. Population genetic theory predicts that selfing species should have low rates of heterozygosity (few heterozygous sites in the genome). Compare your Arabidopsis GenomeScope kmer-profile to the k-mer profile and GenomeScope output statistics for date palm, an obligately outcrossing species below:

![Date Palm K-mer Profile](datepalm_GenomeScope_profiles.png) 
Fig. 1. GenomeScope kmer profile for date palm ('Al Ain male') shown for raw coverage (left) and log-scaled coverage (right). Date palm GenomeScope summary statistics are shown at the top of the graphs.

Q3.3a. Does the kmer-profiling estimates of heterozygosity (the proportion of heterozygous sites) for Arabidopsis and date palm match this expectation? [ 1 point ].

[**Answer**]{.underline}: the proportion of heterozygous sites for Arabidopsis and date palm match the expectation:

- the heterozygosity for Arabidopsis: 0.0994%;

- the heterozygosity for data palm: 1.25%.

Q3.3b. From left to right, which of the peaks in the output profile plot for date palm is the homozygous peak, the heterozygous peak, the error peak? Why is the heterozygous peak expected to be half of the homozygous peak (in terms of number of occurences or "Coverage")? [ 1 points ]
[**Answer**]{.underline}:The bimodal distribution typical of a diploid heterozygous genome. from left to right, the peaks are error peak (area under orange-line), heterozygous peak, and homozygous peak. 

the coverage of heterozygous peak is expected to be half of the homozygous peak because genome heterozygosity is haploid whereas homozygous genome peak is diploid.

Q3.4. A published assembly for date palm has a haploid length of 772 Mb, but may be larger (Hazzouri et al. 2019). The Arabidopsis thaliana assembly is much smaller (\~135 Mb). Summarize the haploid genome size estimates of GenomeScope for these two species? Then speculate as to why GenomeScope estimates might differ from the genome assembly sizes for both species [ 1 point ].

[**Answer**]{.underline}: the genome size estimated was about 355Mb for Date palm and 124Mb for Arabidopsis thaliana. The reason that the estimated genome size different for these two species because their differences in genome heterogeneity. In GenomeScope, the total genome size is
estimated by normalizing the observed k-mer frequencies to the average coverage value for homozygous sequences, excluding likely sequencing errors. Highly heterogygous genome - Date palm is considered having high rate of error sequence and thus part of the sequences could be removed from total genome size. Arabidopsis thaliana genome is homozygous, and less "error" sequences be remove by the program.

Q3.5. Hypothetically-speaking, how would you expect the k-mer profile to differ between two fastqs from the same sample with a 2-fold difference in sequencing error rate? [ 1 point ]

[**Answer**]{.underline}:Hypothetically-speaking,the genome with higher error rate would have lower genome size with higher error peak. Given that some error may be considered heterozygosity, it could have higher heterozygous peak and lower homozygous peak compared to low error rate fastq. 


